lib/bolt/puppetfile/forge_spec.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'semantic_puppet'
+require 'bolt/error'
+
+# This class represents a Forge module specification.
+#
+module Bolt
+  class Puppetfile
+    class ForgeSpec
+      NAME_REGEX    = %r{\A[a-z][a-z0-9_]*[-/](?<name>[a-z][a-z0-9_]*)\z}.freeze
+      REQUIRED_KEYS = %w[name].freeze
+      KNOWN_KEYS    = %w[name version_requirement].freeze
+
+      attr_reader :full_name, :name, :type, :version_requirement
+
+      def initialize(name, version_requirement = nil)
+        @full_name, @name    = parse_name(name)
+        @version_requirement = parse_version_requirement(version_requirement)
+        @type                = :forge
+      end
+
+      def self.from_hash(hash)
+        new(hash['name'], hash['version_requirement'])
+      end
+
+      def self.implements?(hash)
+        (hash.keys & REQUIRED_KEYS) == REQUIRED_KEYS && (hash.keys - KNOWN_KEYS).empty?
```

Comment:
This will also verify the order which I don't think we want, since:
```
modules:
  - ref: 'main'
    git: https://github.com/mycool/module
```
will error. I think converting to sets an comparing those is most reliable, but sorting and comparing is also common ([benchmarks](https://stackoverflow.com/a/56739603/2708055)) (ooo or there's a new `difference` function when we upgrade to 2.6, though I'm not sure it's faster just maybe more readable).

Reply:
It also seems like a lot of this is copied from the ForgeSpec class - it seems like we should either have a parent class that can handle both specifications, or tailor each class to it's needs - so this for example wouldn't need to have KNOWN_KEYS